consensus/hotstuff/vote_collector.go

```diff
@@ -43,11 +54,19 @@ type VoteCollector interface {
 	Status() VoteCollectorStatus
 }
 
-// VerifyingVoteCollector must implement the same interface as BlockSigner, such that when eventhandler
-// asks voter to vote for the block, the voter will ask VoteAggregator(via BlockSigner interface)
-// to sign the block, and VoteAggregator will read the VerifyingVoteCollector from the vote collectors
-// map and produce the vote.
-// Note CachingVoteCollector can't create vote, only VerifyingVoteCollector can
-type VerifyingVoteCollector interface {
-	BlockSigner
+type VoteCollector interface {
+	VoteCollectorState
+	// ChangeProcessingStatus changes the VoteCollector's internal processing
+	// status. The operation is implemented as an atomic compare-and-swap, i.e. the
+	// state transition is only executed if VoteCollector's internal state is
+	// equal to `expectedValue`. The return indicates whether the state was updated.
+	// The implementation only allows the transitions
+	//         CachingVotes -> VerifyingVotes
+	//    and                  VerifyingVotes -> Invalid
+	// Error returns:
+	// * nil if the state transition was successfully executed
+	// * ErrDifferentCollectorState if the VoteCollector's state is different than expectedCurrentStatus
+	// * ErrInvalidCollectorStateTransition if the given state transition is impossible
+	// * all other errors are unexpected and potential symptoms of internal bugs or state corruption (fatal)
+	ChangeProcessingStatus(expectedValue, newValue VoteCollectorStatus) error
```

Comment:
Why a generic and error-prone state machine method (e.g. does it make sense to want to switch from `VoteCollectorStatusInvalid`?), rather than one or several specialized one(s) for a voting state machine (e.g. `Flush`: VoteCollectorStatusCaching -> VoteCollectorStatusVerifying)

Reply:
That is the approach that we took in approval collecting, this special interface is intended to be used only by a class which holds collectors and understands how to change state of state machine for everything else `VoteCollectorState` is sufficient.